compiler/src/iree/compiler/DispatchCreation/FoldUnitExtentDims.cpp

```diff
@@ -32,6 +33,146 @@ namespace mlir::iree_compiler::DispatchCreation {
 #define GEN_PASS_DEF_FOLDUNITEXTENTDIMSPASS
 #include "iree/compiler/DispatchCreation/Passes.h.inc"
 
+namespace {
+
+/// Simplify collapse_shape(expand_shape) by removing unneeded unit dimensions
+/// that get expanded and subsequently collapsed.
+struct DropUnitDimsFromCollapseOfExpand
+    : OpRewritePattern<tensor::CollapseShapeOp> {
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(tensor::CollapseShapeOp collapseOp,
+                                PatternRewriter &rewriter) const override {
+    auto expandOp = collapseOp.getSrc().getDefiningOp<tensor::ExpandShapeOp>();
+    if (!expandOp) {
+      return failure();
+    }
+
+    const auto collapseReassoc = collapseOp.getReassociationIndices();
+    ArrayRef<int64_t> interShape = expandOp.getType().getShape();
+    ArrayRef<int64_t> outShape = collapseOp.getType().getShape();
+    SmallVector<int64_t> interToOutMap(expandOp.getType().getRank());
+    llvm::SmallDenseSet<int64_t> toDrop;
+    for (const auto &[outDim, indicies] : llvm::enumerate(collapseReassoc)) {
+      for (auto [innerIdx, inDim] : llvm::enumerate(indicies)) {
+        // Can't drop this dim if it isnt statically 1 or if it isn't being
+        // combined with any other dimensions.
+        if (indicies.size() == 1 || interShape[inDim] != 1) {
+          continue;
+        }
+
+        // If we are collapsing multiple unit dims together, at least 1 must be
+        // kept (prefer the first).
+        if (outShape[outDim] == 1 && innerIdx != 0) {
```

Comment:
I think I understand what you are trying to do, but I dont follow the logic that much. Maybe some comments would help?